src/psyclone/dynamo0p3.py

```diff
@@ -9020,31 +9052,32 @@ def scalars(self):
 
 
 class DynKernelArgument(KernelArgument):
-    ''' Provides information about individual Dynamo kernel call
-    arguments as specified by the kernel argument metadata. '''
+    '''
+    This class provides information about individual LFRic kernel call
+    arguments as specified by the kernel argument metadata.
+
+    :param kernel_args: object encapsulating all arguments to the \
+                        kernel call.
+    :type kernel_args: :py:class:`psyclone.dynamo0p3.DynKernelArguments`
+    :param arg_meta_data: information obtained from the meta-data for \
+                          this kernel argument.
+    :type arg_meta_data: :py:class:`psyclone.dynamo0p3.LFRicArgDescriptor`
+    :param arg_info: information on how this argument is specified in \
+                     the Algorithm layer.
+    :type arg_info: :py:class:`psyclone.parse.algorithm.Arg`
+    :param call: the kernel object with which this argument is associated.
+    :type call: :py:class:`psyclone.dynamo0p3.DynKern`
 
+    '''
     def __init__(self, kernel_args, arg_meta_data, arg_info, call):
-        '''
-        :param kernel_args: Object encapsulating all arguments to the
-                            kernel call
-        :type kernel_args: :py:class:`psyclone.dynamo0p3.DynKernelArguments`
-        :param arg_meta_data: Information obtained from the meta-data for
-                              this kernel argument
-        :type arg_meta_data: \
-            :py:class:`psyclone.domain.lfric.LFRicArgDescriptor`
-        :param arg_info: Information on how this argument is specified in the
-                         Algorithm layer
-        :type arg_info: :py:class:`psyclone.parse.algorithm.Arg`
-        :param call: The kernel object with which this argument is associated
-        :type call: :py:class:`psyclone.dynamo0p3.DynKern`
-        '''
         KernelArgument.__init__(self, arg_meta_data, arg_info, call)
         # Keep a reference to DynKernelArguments object that contains
         # this argument. This permits us to manage name-mangling for
         # any-space function spaces.
         self._kernel_args = kernel_args
         self._vector_size = arg_meta_data.vector_size
         self._type = arg_meta_data.type
+        self._intrinsic_type = arg_meta_data.data_type
```

Comment:
I suspect (but am not sure) that this is the point at which we should map from metadata information into DSL information. i.e. what the intrinsic type of the object is. I think the type should be one of `psyir.symbols.ScalarType.Intrinsic.INTEGER/REAL/...` and should be stored in some map (dict). Ultimately this could come from the configuration file but obviously not yet :-)
At the moment, the mapping from metadata to intrinsic type is done in LFRicArgDescriptor which is not a bad place but my feeling is that that class should deal purely with the kernel metadata. (Possibly it and those related to it should be renamed to e.g. LFRicArgMetadata to make this clear.) The current class brings together the information from the Algorithm and the kernel metadata and is therefore maybe a more natural place to introduce this type information.

Reply:
Yes, I thought about it, too. My initial implementation in #774 was to just leave `LFRicArgDescriptor.data_type` as metadata. The mapping from descriptor to intrinsic type would happen in the `DynKernelArgument` class.
For instance, descriptor `data_type` for e.g. `(GH_FIELD, GH_READ, W2)` would be `GH_REAL` and `(GH_SCALAR, GH_INTEGER, GH_READ)` would be `GH_INTEGER`. Argument `intrinsic_type` would then be determined by some mapping from that.

The solution implemented here was to a) put something in place until we have metadata for indicating the intrinsic type and b) limit the simultaneous support for the current and the new metadata just to the `LFRicArgDescriptor` class. For instance, the support for fields would be something like
```
if current metadata (e.g. (GH_FIELD, GH_READ, W2)) then
    data_type = `real`
elif new metadata (e.g. (GH_FIELD, GH_REAL, GH_READ, W2)) then
    data_type = descriptor_from_metadata(arg_type.args[1])
```
and for scalars
```
if current metadata (e.g. (GH_INTEGER, GH_READ)) then
    data_type = descriptor_from_metadata(arg_type.args[0])
elif new metadata (e.g. (GH_SCALAR, GH_INTEGER, GH_READ)) then
    data_type = descriptor_from_metadata(arg_type.args[1])
```

A  way out of this conundrum would be to simply define `data_type` as `GH_REAL` or `GH_INTEGER` in the descriptor class in this PR, i.e.
```
if current metadata (e.g. (GH_FIELD, GH_READ, W2)) then
    data_type = GH_REAL
elif new metadata (e.g. (GH_FIELD, GH_REAL, GH_READ, W2)) then
    data_type = descriptor_from_metadata(arg_type.args[1])
```
and
```
if current metadata (e.g. (GH_INTEGER, GH_READ)) then
    data_type = arg_type.args[0]
elif new metadata (e.g. (GH_SCALAR, GH_INTEGER, GH_READ)) then
    data_type = arg_type.args[1]
```
and introduce the mapping to intrinsic type somewhere else (as said above, my initial idea was `DynKernelArgument` class but I am open to suggestions). What do you think? 
Of course, that would mean essentially duplicating the current scalar names but we would do that anyway if we want to support both

Reply:
I decided to implement the second solution where the `LFRicArgDescriptor.data_type` is one of `LFRicArgDescriptor.VALID_ARG_DATA_TYPES = ["gh_real", "gh_integer"]`. The scalar `data_type` is now simply its name but that will change in #774. Also, #774 will establish a check whether `data_type` metadata descriptor is one of `LFRicArgDescriptor.VALID_ARG_DATA_TYPES`. Fields and operators now have `gh_real` but that will be extended in #817. 

`DynKernelArgument.intrinsic_type` is now set via a `dynamo0p3.MAPPING_DATA_TYPES` mapping with the `InternalError` for a wrong key.